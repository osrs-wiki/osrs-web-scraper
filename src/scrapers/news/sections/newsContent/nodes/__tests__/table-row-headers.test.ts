import {
  MediaWikiBuilder,
} from "@osrs-wiki/mediawiki-builder";
import parse from "node-html-parser";

import tableParser from "../table";

describe("table node - row headers", () => {
  test("A table with row headers (<th> in table body) should render correctly", () => {
    const root = parse(`
      <table>
        <tbody>
          <tr>
            <th></th>
            <th>EU</th>
            <th>UK</th>
            <th>US East</th>
          </tr>
          <tr>
            <th>F2P</th>
            <td>456, 455</td>
            <td>498, 499</td>
            <td>417, 648</td>
          </tr>
          <tr>
            <th>Members</th>
            <td>335, 384</td>
            <td>380, 382</td>
            <td>629, 630</td>
          </tr>
        </tbody>
      </table>
    `);
    const builder = new MediaWikiBuilder();
    const tableNode = root.querySelector("table");
    builder.addContents([tableParser(tableNode)].flat());
    const result = builder.build();
    
    // Check that the result has the correct header markers
    expect(result).toContain("! \n! EU");
    expect(result).toContain("! F2P\n| 456, 455");
    expect(result).toContain("! Members\n| 335, 384");
    
    expect(result).toMatchSnapshot();
  });

  test("A table with all data cells should render correctly", () => {
    const root = parse(`
      <table>
        <tbody>
          <tr>
            <td>Row 1 Col 1</td>
            <td>Row 1 Col 2</td>
          </tr>
          <tr>
            <td>Row 2 Col 1</td>
            <td>Row 2 Col 2</td>
          </tr>
        </tbody>
      </table>
    `);
    const builder = new MediaWikiBuilder();
    const tableNode = root.querySelector("table");
    builder.addContents([tableParser(tableNode)].flat());
    const result = builder.build();
    
    // First row becomes header row
    expect(result).toContain("! Row 1 Col 1");
    expect(result).toContain("! Row 1 Col 2");
    // Second row should be data cells
    expect(result).toContain("| Row 2 Col 1");
    expect(result).toContain("| Row 2 Col 2");
    
    expect(result).toMatchSnapshot();
  });

  test("A table with mixed th and td in rows should render correctly", () => {
    const root = parse(`
      <table>
        <tbody>
          <tr>
            <th>Header 1</th>
            <th>Header 2</th>
            <th>Header 3</th>
          </tr>
          <tr>
            <th>Row Header</th>
            <td>Data 1</td>
            <td>Data 2</td>
          </tr>
        </tbody>
      </table>
    `);
    const builder = new MediaWikiBuilder();
    const tableNode = root.querySelector("table");
    builder.addContents([tableParser(tableNode)].flat());
    const result = builder.build();
    
    // All cells in first row should be headers
    expect(result).toContain("! Header 1");
    expect(result).toContain("! Header 2");
    expect(result).toContain("! Header 3");
    
    // Second row should have mixed cell types
    expect(result).toContain("! Row Header");
    expect(result).toContain("| Data 1");
    expect(result).toContain("| Data 2");
    
    expect(result).toMatchSnapshot();
  });

  test("A table with thead and tbody row headers should render correctly", () => {
    const root = parse(`
      <table>
        <thead>
          <tr>
            <th>Column 1</th>
            <th>Column 2</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th>Row 1</th>
            <td>Value 1</td>
          </tr>
          <tr>
            <th>Row 2</th>
            <td>Value 2</td>
          </tr>
        </tbody>
      </table>
    `);
    const builder = new MediaWikiBuilder();
    const tableNode = root.querySelector("table");
    builder.addContents([tableParser(tableNode)].flat());
    const result = builder.build();
    
    // thead cells should be headers
    expect(result).toContain("! Column 1");
    expect(result).toContain("! Column 2");
    
    // tbody row headers should be headers
    expect(result).toContain("! Row 1");
    expect(result).toContain("! Row 2");
    
    // tbody data cells should be data
    expect(result).toContain("| Value 1");
    expect(result).toContain("| Value 2");
    
    expect(result).toMatchSnapshot();
  });
});
