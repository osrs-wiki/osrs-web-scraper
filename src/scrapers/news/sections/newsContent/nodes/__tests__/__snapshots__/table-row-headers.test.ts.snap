// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table node - row headers A table with all data cells should render correctly 1`] = `
"{| class=\\"wikitable\\" style=\\"text-align: center;\\"
|-
! Row 1 Col 1
! Row 1 Col 2
|-
| Row 2 Col 1
| Row 2 Col 2
|}
"
`;

exports[`table node - row headers A table with mixed th and td in rows should render correctly 1`] = `
"{| class=\\"wikitable\\" style=\\"text-align: center;\\"
|-
! Header 1
! Header 2
! Header 3
|-
! Row Header
| Data 1
| Data 2
|}
"
`;

exports[`table node - row headers A table with row headers (<th> in table body) should render correctly 1`] = `
"{| class=\\"wikitable\\" style=\\"text-align: center;\\"
|-
! 
! EU
! UK
! US East
|-
! F2P
| 456, 455
| 498, 499
| 417, 648
|-
! Members
| 335, 384
| 380, 382
| 629, 630
|}
"
`;

exports[`table node - row headers A table with thead and tbody row headers should render correctly 1`] = `
"{| class=\\"wikitable\\" style=\\"text-align: center;\\"
|-
! Column 1
! Column 2
|-
! Row 1
| Value 1
|-
! Row 2
| Value 2
|}
"
`;
